fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -861,6 +865,7 @@ std::set<Tag> CommitBatchContext::getWrittenTagsPreResolution() {
 		toCommit.storeRandomRouterTag();
 		transactionTags.insert(toCommit.savedRandomRouterTag.get());
 	}
+	lastShardMove = pProxyCommitData->lastShardMove;
 
```

Comment:
Isn't there a timing issue here? The last shard move version would need to be obtained at the time the transaction gets started/the time the tags get collected? I think it should be at the point where we start accessing the shard metadata in this transaction? Else, the last shard move version could get updated in this window and we might not report the correct version to the resolver?

Reply:
The PR description depicts the case where we cannot obtain the last shard move version at the time transaction A starts, because another transaction B may be running in parallel, and A does not learn of the shard move until after B returns. By then A has already chosen its tags using the old shard map. i.e. transactions A and B run in parallel.

Reply:
>  last shard move version would need to be obtained at the time the transaction gets started/the time the tags get collected

This function is where the CP collects the tags for each key. `lastShardMove` is the version last known to have modified the shard map. The CP needs to report to the resolver the version it used in this function. If it the shard map is modified after returning from this function, the resolver's `lastShardMove` will advance beyond the value used in this function, which will indicate the batch much be sent to all logs.

Reply:
It appears there is more complexity here. I see that this function is using cached data to figure out the tags - are we sure that the cached metadata reflects the shard metadata changes done till the "shard move" version that we are collecting here?

Reply:
As far as I know, yes. The RV processes in version order, and any change to the shard map must be done via a transaction that first goes through the resolver before making the change post resolution.

Reply:
Agreed. Posted a comment about this below.